src/Workspaces/Remote/Core/Serialization/ImmutableCollectionMessagePackResolver.cs

```diff
@@ -0,0 +1,394 @@
+            { typeof(IImmutableSet<>), typeof(InterfaceImmutableSetFormatter<>) },
+            { typeof(IImmutableStack<>), typeof(InterfaceImmutableStackFormatter<>) },
+        };
+
+        internal static object GetFormatter(Type t)
+        {
+            var ti = t.GetTypeInfo();
+
+            if (ti.IsGenericType)
+            {
+                var genericType = ti.GetGenericTypeDefinition();
+                var genericTypeInfo = genericType.GetTypeInfo();
+                var isNullable = genericTypeInfo.IsGenericType && genericTypeInfo.GetGenericTypeDefinition() == typeof(Nullable<>);
+                var nullableElementType = isNullable ? ti.GenericTypeArguments[0] : null;
+
+                if (s_formatterMap.TryGetValue(genericType, out var formatterType))
+                {
+                    return CreateInstance(formatterType, ti.GenericTypeArguments);
+                }
+
+                if (isNullable && nullableElementType.IsConstructedGenericType && nullableElementType.GetGenericTypeDefinition() == typeof(ImmutableArray<>))
+                {
+                    return CreateInstance(typeof(NullableFormatter<>), new[] { nullableElementType });
+                }
+            }
+
+            return null;
+        }
+
+        private static object CreateInstance(Type genericType, Type[] genericTypeArguments, params object[] arguments)
+            => Activator.CreateInstance(genericType.MakeGenericType(genericTypeArguments), arguments);
+
+        // ImmutableArray<T>.Enumerator is 'not' IEnumerator<T>, can't use abstraction layer.
+        internal sealed class ImmutableArrayFormatter<T> : IMessagePackFormatter<ImmutableArray<T>>
+        {
+            public void Serialize(ref MessagePackWriter writer, ImmutableArray<T> value, MessagePackSerializerOptions options)
+            {
+                if (value.IsDefault)
+                {
+                    writer.WriteNil();
+                }
+                else
+                {
+                    IMessagePackFormatter<T> formatter = options.Resolver.GetFormatterWithVerify<T>();
+
+                    writer.WriteArrayHeader(value.Length);
+
+                    foreach (T item in value)
+                    {
+                        formatter.Serialize(ref writer, item, options);
+                    }
+                }
+            }
+
+            public ImmutableArray<T> Deserialize(ref MessagePackReader reader, MessagePackSerializerOptions options)
+            {
+                if (reader.TryReadNil())
+                {
+                    return default;
+                }
+                else
+                {
+                    IMessagePackFormatter<T> formatter = options.Resolver.GetFormatterWithVerify<T>();
+
+                    var len = reader.ReadArrayHeader();
+
+                    ImmutableArray<T>.Builder builder = ImmutableArray.CreateBuilder<T>(len);
+                    options.Security.DepthStep(ref reader);
```

Comment:
interesting.  what ist he default set to?

Reply:
See the comment at the beginning of the file.